### DISCUSSION TITLE
Running python script in C#, function cannot be called twice

### DISCUSSION THREAD

**Discussion by Jiew23** [pythonnet/pythonnet#2051] [Q&A]
### Environment

-   Pythonnet version: 3.0.1
-   Python version: 3.9.12
-   Operating System: All
-   .NET Runtime: All

### Details

I am writing a function in C# application which executes the python script. I realised that the C# function that runs python script can only be called once for each C# project run. If I need to run another python script. I will have to terminate the C# project and restart. 

My function is as below.

```python
     public PyObject ExecutePythonScript(string pythonScript)
    {
        PyObject result;
        PythonEngine.Initialize();
        using (Py.GIL())
        {
            using (var scope = Py.CreateScope())
            {
                dynamic a_library = Py.Import("a_library");
                scope.Exec("import a_library");
                scope.Exec(pythonScript);
                result = scope.Eval("result");
            }
        }
        return result;
   }
```

For each C# project restart, the function runs with no problems for the first time. But if try to run it twice, it will stuck at using (Py.GIL()).

Does anyone know how to terminate the python execution without terminate the C# project so I could run this function again in the same project run? Much appreciated!

**Comment by tnsc2022** [comment]
Maybe you should add `PythonEngine.Shutdown();` after `using {}`.
Or change `Runtime.PythonDLL = pythonDll;` to `Environment.SetEnvironmentVariable("PYTHONNET_PYDLL", pythonDll, target);`

**Reply by Jiew23** [reply]
Hi tnsc2022,

I tried PythonEngine.Shutdown() but it throws an error when trying to read the variable returned from the python function. 

The error is as below:

System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

The Environment.SetEnvironment also did not help. 

Do you have any other recommendations on what we could do to solve this problem? Thank you so much.
